src/System.Security.Cryptography.Primitives/src/System/Security/Cryptography/CryptoStream.cs

```diff
@@ -128,7 +128,7 @@ public void FlushFinalBlock()
                     innerCryptoStream.FlushFinalBlock();
                 }
             }
-            else
+            else if (_canWrite)
```

Comment:
A FileStream opened in read mode still has semantic behaviors on Flush, and not calling it could have observable consequences to caller code (and that was simply the first kind of stream I looked at).  I don't know that there's a safe way to make this change.

Reply:
So, basically, you are telling that my issue (`CryptoStream` throwing exceptions in `Dispose`) is a bug in Kestrel server that doesn't allow to call `Flush` on read-only stream, right?

Reply:
Yep, sorry.  If they can't/won't fix it you could add yet another Stream as a wrapper over it, which just intercepts the call to Flush and doesn't pass it on.

I checked the .NET Framework 2.0 source base, and CryptoStream definitely had this behavior back that far.